ADAL/src/request/ADAuthenticationRequest.m

```diff
@@ -268,4 +268,14 @@ - (NSUUID*)correlationId
     return _correlationId;
 }
 
+- (BOOL)takeUserInterationLock
+{
+    return OSAtomicCompareAndSwapInt( 0, 1, &sInteractionInProgress);
```

Comment:
I tried to use NSLock, but I found an interesting thing: the reference says 
"When sending an unlock message to an NSLock object, you must be sure that message is sent from the same thread that sent the initial lock message. Unlocking a lock from a different thread can result in undefined behavior."

We may need to release a lock in a completion block, where the thread executing the block is uncertain :P 

For "dispatch primitive", do you mean dispatch_once? :)